src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte512Vector.java

```
@@ -949,39 +950,103 @@ public ByteSpecies vspecies() {
+        }
+
+        @ForceInline
+        @Override
+        public final Byte512Shuffle rearrange(VectorShuffle<Byte> shuffle) {
+            Byte512Shuffle concreteShuffle = (Byte512Shuffle) shuffle;
+            return (Byte512Shuffle) toBitsVector().rearrange(concreteShuffle)
+                    .toShuffle(vspecies(), false);
+        }
+
+        @ForceInline
+        @Override
+        public final Byte512Shuffle wrapIndexes() {
+            Byte512Vector v = toBitsVector();
+            if ((length() & (length() - 1)) == 0) {
+                v = (Byte512Vector) v.lanewise(VectorOperators.AND, length() - 1);
+            } else {
+                v = (Byte512Vector) v.blend(v.lanewise(VectorOperators.ADD, length()),
+                            v.compare(VectorOperators.LT, 0));
+            }
+            return (Byte512Shuffle) v.toShuffle(vspecies(), false);
+        }
+
+        private static byte[] prepare(int[] indices, int offset) {
+            byte[] a = new byte[VLENGTH];
+            for (int i = 0; i < VLENGTH; i++) {
+                int si = indices[offset + i];
+                si = partiallyWrapIndex(si, VLENGTH);
+                a[i] = (byte)si;
+            }
+            return a;
+        }
+
+        private static byte[] prepare(IntUnaryOperator f) {
+            byte[] a = new byte[VLENGTH];
+            for (int i = 0; i < VLENGTH; i++) {
+                int si = f.applyAsInt(i);
+                si = partiallyWrapIndex(si, VLENGTH);
+                a[i] = (byte)si;
+            }
+            return a;
+        }
+
+        private static boolean indicesInRange(byte[] indices) {
+            int length = indices.length;
+            for (byte si : indices) {
+                if (si >= (byte)length || si < (byte)(-length)) {
+                    String msg = ("index "+si+"out of range ["+length+"] in "+
+                                  java.util.Arrays.toString(indices));
+                    throw new AssertionError(msg);
```

Why not directly throw IndexOutOfBoundsException here?

---

This is called in an `assert` so I think throwing `AssertionError` seems more reasonable, the original implementation also throws `AssertionError`

---

It seems you want to throw an error here and not just use an assert statement that reports any assertion failures with -ea flag.  This is a newly added routine, why don't you simply return a false and let the assertion invoking this routine do the rest?

---

This routine is moved from `AbstractShuffle` where it was implemented when all shuffle types share the same constructor  there. Throwing here allows easier construction of the error message.